### PR TITLE
more specific type hints, value errors in Multicall

### LINKED ISSUE #339: ValueError...'unsupported block number 21036641'
This is a new error message / issue that stops telliot from running occurred while reporting to Arbitrum.

start command: `telliot-feeds -a arb1 report -qt eth-usd-legacy -gl 5000000 -gp 100 -p YOLO -wp 480`

full traceback:
[unsupported block number.txt](https://github.com/tellor-io/telliot-feeds/files/9402636/unsupported.block.number.txt)

### PR DESCRIPTION
**Pull request by tallywiesenberg**
### Summary

Multicalls in this `reporter_autopay_utils` did not have error handling, so I added error handling for `ValueError`s for RPC errors and `web3.exceptions.ContractLogicError` for contract reversions.

I also added more specific type hints to the multicall functions.
### Steps Taken to QA Changes

- Ran pytest, removed expected raise of `ContractLogicError`
- Ran mypy to improve type hints

### Checklist

This pull request is:

- [ ] A documentation error, docs update, or typographical error fix
	- No tests or issue needed
- [x] A code fix
	- Please reference the related issue by including "Closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a bug report issue 
	- Please include tests. Fixes without tests will not be accepted unless it's related to the documentation only.
    - Please make sure docs are updated if need be
- [ ] A feature implementation
	- Please reference the related issue by including "Closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a feature enhancement issue 
	- Please include tests
    - Please make sure docs updates are both thorough and easy to reproduce by somebody with limited knowledge of the feature that you are submitting

Closes #339
**Happy engineering!**

### CODE REVIEW

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 1]
```diff
@@ -122,14 +123,25 @@ async def get_current_feeds(self, require_success: bool = False) -> Any:
                     )
                 )
         multi_call = Multicall(calls=calls, _w3=self.w3, require_success=require_success)
-        data = await multi_call.coroutine()
+        try:
+            data: Dict[str, Any] = await multi_call.coroutine()
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:127]: you don't need to define the types for `data`

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 2]
```diff
@@ -122,14 +123,25 @@ async def get_current_feeds(self, require_success: bool = False) -> Any:
...
+        except ValueError as e:
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:128]: You're expecting `ValueError`, but your logging messages say "KeyError"

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 3]
```diff
@@ -122,14 +123,25 @@ async def get_current_feeds(self, require_success: bool = False) -> Any:
...
+                msg = f"Error in multicall request, KeyError {e}"
+                logger.warning(msg)
+                return None
         # remove status boolean thats useless here
         try:
-            data.pop("disregard_boolean")
+            _ = data.pop("disregard_boolean")
+            return data
         except KeyError as e:
             msg = f"No feeds returned by multicall, KeyError: {e}"
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:142]: It's a little unclear that this this key not being in the dict means no feeds were returned, maybe add a comment why. I know this is unrelated to the issue, but this is as good an opportunity as any..

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 4]
```diff
@@ -154,13 +166,16 @@ async def get_feed_details(self, require_success: bool = False) -> Any:
         # separate items from current feeds
         # create dict of tag and feed_id in current_feeds
         tags_with_feed_ids = {
-            tag: feed_id for tag, feed_id in current_feeds.items() if type(tag) != tuple if len(current_feeds[tag]) > 0
+            tag: feed_id
+            for tag, feed_id in current_feeds.items()
+            if isinstance(tag, tuple)
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:171]: using
```python
blah for blah in blah if all(condition1, condition1)
```
would be clearer than this dict comprehension with two if statements..
plz change

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 1 above)
Comment (oraclown) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:127]: nvm, if you want to make types clearer, idk why we shouldn't do that

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 5]
```diff
@@ -211,18 +226,28 @@ def _to_list(_: bool, val: Any) -> List[Any]:
         ]
         calls = get_data_feed_call + get_current_values_call + get_timestampby_query_id_n_idx_call
         multi_call = Multicall(calls=calls, _w3=self.w3, require_success=require_success)
-        feed_details = await multi_call.coroutine()
-
-        return feed_details
+        try:
+            feed_details = await multi_call.coroutine()
+            return feed_details
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:231]: I think you can just `return await blah`, no?

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 6]
```diff
@@ -211,18 +226,28 @@ def _to_list(_: bool, val: Any) -> List[Any]:
...
+        except ValueError as e:
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:232]: Same thing re your logging msg here. Also, seems like it's repeating some logic above, so let's refactor this into a single function if possible.

File: src/telliot_feeds/reporters/reporter_autopay_utils.py [hunk 7]
```diff
@@ -264,11 +289,20 @@ async def reward_claim_status(self, require_success: bool = False) -> Any:
                             )
 
         multi_call = Multicall(calls=reward_claimed_status_call, _w3=self.w3, require_success=require_success)
```
Comment (oraclown) [inline src/telliot_feeds/reporters/reporter_autopay_utils.py:291]: All these `mutli_call` reused var names should be made specific to each one, so the code's easier to read/understand. Like `multicall_claim_rewards` or something

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 5 above)
Comment (oraclown) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:231]: Same thing a couple times below in your code additions

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 6 above)
Comment (oraclown) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:232]: Same thing a couple times in the code further down

File: tests/reporters/test_autopay_multicalls.py [hunk 8]
```diff
@@ -29,10 +29,6 @@ async def test_get_current_tips(setup_autopay_call):
     tips = await calls.get_current_tip(require_success=False)
     assert tips["eth-usd-spot"] is None
 
-    with pytest.raises(ContractLogicError):
```
Comment (oraclown) [inline tests/reporters/test_autopay_multicalls.py:32]: Why was this taken out?

**Review by oraclown — CHANGES_REQUESTED** [review]
@tallywiesenberg thanks for this! There's some refactoring needed, and I added a few questions. Also, some of the requests aren't related to your additions exactly, but the best time to make them has past haha (since I missed giving these points on an earlier review), so ya sry.

We also need a test reproducing the issue-specific failure and demonstrating your added logging behavior for this.

**Comment by oraclown** [comment]
The type-checking CI run failed btw. Seems like an ez fix

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 7 above)
Comment (akremstudy) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:291]: I agree thats my bad, I'll keep this in mind.  Also just a heads up I'll be making a lot of changes on this file for 360 updates and I'll be more conscious about the var names.

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 1 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:127]: yeah `mypy` wanted it, but yes i think it also increases code readbility when interacting with outside libraries

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 2 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:128]: whoops!

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 4 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:171]: indeed

File: src/telliot_feeds/reporters/reporter_autopay_utils.py (hunk 5 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/reporter_autopay_utils.py:231]: yes

File: tests/reporters/test_autopay_multicalls.py (hunk 8 above)
Comment (tallywiesenberg) [reply tests/reporters/test_autopay_multicalls.py:32]: before this function raised this error, instead now I catch it

**Review by oraclown — CHANGES_REQUESTED** [review]
Nice changes @tallywiesenberg . Still need test reproducing original bad behavior highlighted in the issue. The tests should also reflect the logging changes in response to those errors that you've made

**Comment by tallywiesenberg** [comment]
thanks @oraclown unit test w/ error handling added !

**Review by oraclown — APPROVED** [review]
Awesome @tallywiesenberg , thanks!

**Pull request merged**